service/frontend/adminHandler.go

```diff
@@ -118,21 +136,180 @@ func (adh *AdminHandler) DescribeWorkflowExecution(ctx context.Context, request
+	}
+
+	pageSize := int(request.GetMaximumPageSize())
+	if pageSize < 0 {
+		return nil, &gen.BadRequestError{Message: "Invalid PageSize."}
+	}
+
+	var token *getHistoryContinuationToken
+	// initialize or validate the token
+	// token will be used as a source of truth
+	if request.NextPageToken != nil {
+		token, err = deserializeHistoryToken(request.NextPageToken)
+		if err != nil {
+			return nil, err
+		}
+
+		if execution.GetRunId() != token.RunID ||
+			// we guarantee to use the first event ID provided in the request
+			request.GetFirstEventId() != token.FirstEventID ||
+			// the next event ID in the request must be <= next event ID from mutable state, when initialized
+			// so as long as customer do not change next event ID during pagination,
+			// next event ID in the token <= next event ID in the request.
+			request.GetNextEventId() < token.NextEventID {
+			return nil, &gen.BadRequestError{Message: "Invalid pagination token."}
+		}
+
+		// for the rest variables in the token, since we do not do hmac,
+		// the only thing can be done is to trust the token:
+		// IsWorkflowRunning: not used
+		// TransientDecision: not used
+		// PersistenceToken: trust
+		// EventStoreVersion: trust
+		// ReplicationInfo: trust
+
+	} else {
+		firstEventID := request.GetFirstEventId()
+		nextEventID := request.GetNextEventId()
+		if firstEventID < 0 || firstEventID > nextEventID {
+			return nil, &gen.BadRequestError{Message: "Invalid FirstEventID && NextEventID combination."}
+		}
+
+		response, err := adh.history.GetMutableState(ctx, &h.GetMutableStateRequest{
+			DomainUUID:          common.StringPtr(domainID),
+			Execution:           execution,
+			ExpectedNextEventId: common.Int64Ptr(common.FirstEventID), // common.FirstEventID means no long poll
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		// check if the input next event ID is > actual next event ID in the mutable state
+		// since we should not leak invalid events
+		if nextEventID > response.GetNextEventId() {
+			nextEventID = response.GetNextEventId()
+		}
+		token = &getHistoryContinuationToken{
+			RunID:             execution.GetRunId(),
+			BranchToken:       response.BranchToken,
+			FirstEventID:      firstEventID,
+			NextEventID:       nextEventID,
+			PersistenceToken:  nil, // this is the initialized value
+			EventStoreVersion: response.GetEventStoreVersion(),
+			ReplicationInfo:   response.ReplicationInfo,
+		}
+	}
+
+	if token.FirstEventID >= token.NextEventID {
+		return &admin.GetWorkflowExecutionRawHistoryResponse{
+			HistoryBatches:    []*gen.DataBlob{},
+			ReplicationInfo:   token.ReplicationInfo,
+			EventStoreVersion: common.Int32Ptr(token.EventStoreVersion),
+			NextPageToken:     nil, // no further pagination
+		}, nil
+	}
+
+	// TODO need to deal with transient decision if to be used by client getting history
+	var historyBatches []*gen.History
+	_, historyBatches, token.PersistenceToken, _, err = historyService.PaginateHistory(
```

Comment:
Can you also emit history size metric with GetWorkflowExecutionRawHistory scope.

Reply:
sure